### PR TITLE
Extend `ActiveRecord::FinderMethods#find` with support for composite pk values

### PR DESCRIPTION
**Pull request by nvasilevski**
`ActiveRecord::FinderMethods#find` now supports passing sets of composite primary key values like:

```ruby
Cpk::Book.find([1, 1])
Cpk::Book.find([[1, 1]])
Cpk::Book.find([1, 1], [1, 2])
Cpk::Book.find([[1, 1], [1, 2]])
```

and treats values as values of the composite primary key columns but only for models with the `primary_key` being an `Array`.

The `find` method has always had a signature that accepts an "identifier" like:
```ruby
Model.find(ID)
Model.find([ID])
Model.find(ID, ID)
Model.find([ID, ID])
```
where `ID` is what identifies a model. It wasn't particularly clear as long as Rails only supported single-column identifies. Though with introduction of the composite primary keys support the meaning of ID as an identifier has extended - now a model can be identified by a set of values and we need to extend `find`'s signature to support that.

### Implementation details

The behavior changes only for models with `primary_key` being an `Array` which led to having several `is_a?(Array)` branches  and a few duplications like `primary_key.zip(values_set)` being a repetitive pattern or the way how we build batch-query relation using `.inject(&:or)` is also repetitive. However at this point I would prefer to introduce duplication to gather as many use-cases as possible and make a refactoring by abstracting these examples into a single concept and having verbose use-cases will only help with finding a proper name and place for the concept.

### CODE REVIEW

**Review by eileencodes — CHANGES_REQUESTED** [review]

File: activerecord/test/cases/finder_test.rb [hunk 1]
```diff
@@ -1781,6 +1782,38 @@ def test_finder_with_offset_string
...
+  test "find with a multiple sets of composite primary key wrapped in an array" do
+    books = [cpk_books(:cpk_great_author_first_book), cpk_books(:cpk_great_author_second_book)]
+
+    assert_equal books.map(&:id).sort, Cpk::Book.find(books.map(&:id)).map(&:id).sort
+  end
+
+  test "find with a multiple sets of composite primary key wrapped in an array ordered" do
+    books = [cpk_books(:cpk_great_author_first_book), cpk_books(:cpk_great_author_second_book)]
+
+    assert_equal books.map(&:id).sort, Cpk::Book.order(author_id: :asc).find(books.map(&:id)).map(&:id).sort
```
[the remark this answers is not in the document]
Comment (matthewd) [reply activerecord/test/cases/finder_test.rb:1814]: Maybe so, but multi-`find` does have a guaranteed return order... seems misleading/confusing to include a definitely-needless `sort`.

File: activerecord/lib/active_record/relation/finder_methods.rb [hunk 2]
```diff
@@ -442,10 +442,21 @@ def using_limitable_reflections?(reflections)
       def find_with_ids(*ids)
         raise UnknownPrimaryKey.new(@klass) if primary_key.nil?
 
-        expects_array = ids.first.kind_of?(Array)
+        expects_array = if primary_key.is_a?(Array)
```
[the remark this answers is not in the document]
Comment (matthewd) [reply activerecord/lib/active_record/relation/finder_methods.rb:445]: I have mixed feelings about encoding `query_constraints_list` == multi-keyed. If we can think of a better name, perhaps a `primary_key_is_an_array?` helper could defer that question?

[the remark this answers is not in the document]
Comment (matthewd) [reply activerecord/lib/active_record/relation/finder_methods.rb:445]: ... or perhaps even `@klass.is_this_a_list_of_ids_rather_than_just_one?(ids.first)`?

[the remark this answers is not in the document]
Comment (nvasilevski) [reply activerecord/lib/active_record/relation/finder_methods.rb:445]: `@klass.composite_primary_key?` sounds reasonable to me. Do you think it would be more reasonable to have a separate PR to introduce `@klass.composite_primary_key?` abstraction? We have more places that still do `primary_key.is_a?(Array)` and I would prefer changing those all at once while also having a separate PR so we can discuss implementation details like whether we should memoize the value or whether the value should be always derived from `primary_key.is_a?(Array)` vs explicitly setting `@composite_primary_key = true` once when we derive it from schema

[the remark this answers is not in the document]
Comment (matthewd) [reply activerecord/lib/active_record/relation/finder_methods.rb:445]: Extract it as a method here, in a separate commit, replacing existing usage as appropriate, but without introducing new logic / caching -- we can consider that part later, but it's fine to do a straight "extract method" here.

File: activerecord/lib/active_record/relation/finder_methods.rb [hunk 3]
```diff
@@ -442,10 +442,21 @@ def using_limitable_reflections?(reflections)
...
+          ids.one? && ids.first.first.is_a?(Array)
```
[the remark this answers is not in the document]
Comment (nvasilevski) [reply activerecord/lib/active_record/relation/finder_methods.rb]: Yep, wasn't necessary, thanks!

**Comment by nvasilevski** [comment]
Addressed comments and extracted `composite_primary_key?` introduction into a separate commit https://github.com/rails/rails/pull/47664/commits/3acf54e4cd412573d7a5ae3d8c5498b701d69f92

File: activerecord/lib/active_record/relation/finder_methods.rb [hunk 4]
```diff
@@ -442,10 +442,21 @@ def using_limitable_reflections?(reflections)
...
+        end
+
         return [] if expects_array && ids.first.empty?
 
-        ids = ids.flatten.compact.uniq
+        ids = if klass.composite_primary_key?
+          expects_array ? ids.first : ids
+        else
+          ids.flatten
+        end
```
Comment (matthewd) [inline activerecord/lib/active_record/relation/finder_methods.rb:457]: This feels hard to follow. The flatten made sense before as an easy way of ignoring `expects_array`... but now I think we can just always respect that:

```suggestion
        ids = expects_array ? ids.first : ids
```

File: activerecord/lib/active_record/relation/finder_methods.rb [hunk 5]
```diff
@@ -469,7 +480,10 @@ def find_one(id)
           MSG
         end
 
-        relation = where(primary_key => id)
+        pk_columns = Array(primary_key)
+        pk_values = Array(id)
+
+        relation = where(pk_columns.zip(pk_values).to_h)
```
Comment (matthewd) [inline activerecord/lib/active_record/relation/finder_methods.rb:486]: This, on the other hand, feels avoidably-allocatey for something that's worth keeping tight. Pending a `where([] => ..)` option, my instinct is that this is probably worth branching on `primary_key.is_a?(Array)`.

File: activerecord/lib/active_record/relation/finder_methods.rb (hunk 4 above)
Comment (nvasilevski) [reply activerecord/lib/active_record/relation/finder_methods.rb:457]: Totally agree, the reason I decided to keep `flatten` for a regular model as it was handling cases like `MyNonCpkModel.find([1,[1,2],[3,4,5],[[[6]]]])` which doesn't look like a supported API though it was working and I decided to keep it to avoid breaking some use-case which I'm not aware of.

But if we all agree that `find` does not support passing non-flattened arrays then `ids.first` will work for all models. I'll make the change

File: activerecord/lib/active_record/relation/finder_methods.rb [hunk 6]
```diff
@@ -506,7 +525,13 @@ def find_some(ids)
       def find_some_ordered(ids)
         ids = ids.slice(offset_value || 0, limit_value || ids.size) || []
 
-        relation = except(:limit, :offset).where(primary_key => ids)
+        relation = except(:limit, :offset)
+        relation = if klass.composite_primary_key?
+          or_scope = ids.map { |values_set| where(primary_key.zip(values_set).to_h) }.inject(&:or)
+          relation.merge(or_scope)
```
Comment (matthewd) [inline activerecord/lib/active_record/relation/finder_methods.rb:531]: merge is scary... is `ids.map { relation.where(..) }.inject(:or)` not a good fit here?

File: activerecord/lib/active_record/relation/finder_methods.rb (hunk 5 above)
Comment (nvasilevski) [reply activerecord/lib/active_record/relation/finder_methods.rb:486]: Actually that's fair, I totally agree with the argument that having a branch for this context clearly reveals an option to get rid of the branch if we ever support `where([] => ..)` syntax. I'll add a branch

File: activerecord/lib/active_record/relation/finder_methods.rb (hunk 6 above)
Comment (nvasilevski) [reply activerecord/lib/active_record/relation/finder_methods.rb]: I clearly missed something in my reasoning as `user_type='admin'` doesn't get duplicated if we just use `relation.where(..)` 

So I simplified it as per your suggestion

Comment (nvasilevski) [reply activerecord/lib/active_record/relation/finder_methods.rb]: I will revisit 
https://github.com/rails/rails/blob/4bd3a108cc245383e8b59ea6ae7937ecb3853587/activerecord/lib/active_record/associations/preloader/association.rb#L40-L41

At least to consider using `scope.and` instead of `scope.merge` but I do remember having some issues with using `scope.where` instead of `scope.klass.where` (clean state)

Eventually I'm hoping we can define an internal helper to build these cross-tenant `OR` statements and reuse it everywhere. But I've been postponing this until we have more use-cases

File: activerecord/lib/active_record/relation/finder_methods.rb [hunk 7]
```diff
@@ -469,7 +476,12 @@ def find_one(id)
           MSG
         end
 
-        relation = where(primary_key => id)
+        relation = if primary_key.is_a?(Array)
```
Comment (eileencodes) [inline activerecord/lib/active_record/relation/finder_methods.rb:479]: This one is not using the `klass.composite_primary_key?`

Comment (nvasilevski) [reply activerecord/lib/active_record/relation/finder_methods.rb]: Ah, lost it in a rebase. Fixed, thanks!

File: activerecord/lib/active_record/relation/finder_methods.rb [hunk 8]
```diff
@@ -469,7 +476,12 @@ def find_one(id)
           MSG
         end
 
-        relation = where(primary_key => id)
+        relation = if klass.composite_primary_key?
+          where(Array(primary_key).zip(Array(id)).to_h)
```
Comment (eileencodes) [inline activerecord/lib/active_record/relation/finder_methods.rb:480]: There doesn't seem to be any tests where these aren't arrays. Is it actually possible for these to not be an array? If not we should remove the `Array` around them to end up with `where(primary_key.zip(id).to_h)`

**Review by eileencodes — COMMENTED** [review]
This is looking good, I have one last question.

File: activerecord/lib/active_record/relation/finder_methods.rb (hunk 8 above)
Comment (nvasilevski) [reply activerecord/lib/active_record/relation/finder_methods.rb]: You are right, this wasn't needed. It was added to handle both single-column and composite pks but then we decided to branch

**Review by eileencodes — APPROVED** [review]

**Comment by nickborromeo** [comment]
👋 we have a use case that passes in `([[ID, ID]])` to find which already seems to work prior to this change. Does this mean that for this format to work going forward it needs to be on a table that has a composite primary key? 

our use case is when we batch queries when running transitions to modify data.

**Comment by nvasilevski** [comment]
>since we were flattening the array before but now we aren't

That's right, this is something we discussed in https://github.com/rails/rails/pull/47664#discussion_r1135888215

Basically the question boils down to whether `.flatten` was unintentionally allowing `find` to accept multi-nested arrays or it was only supposed to unwrap the outer array once like turning `[[1,2,3]]` into `[1,2,3]` If it was done unintentionally then `[[ID, ID]]` format was a non-documented and not supported way of using `find` and it's a caller's responsibility to flatten the argument value before passing it to the method. 

But if multi-nested arrays were intentionally supported I'll be more than happy to bring the flattening back. Perhaps such question has never been raised and now it is the time for us to decide what is the public api for `find` when it comes to passing an array of values.

**Comment by nickborromeo** [comment]
> it was done unintentionally then [[ID, ID]] format was a non-documented and not supported way of using find and it's a caller's responsibility to flatten the argument value before passing it to the method.
>
> But if multi-nested arrays were intentionally supported I'll be more than happy to bring the flattening back. Perhaps such question has never been raised and now it is the time for us to decide what is the public api for find when it comes to passing an array of values

sounds good. I we will handle the flattening on our end for now since there is no decision yet on if this should be supported or not.
